### DISCUSSION TITLE
Make failing for self-created formula

### DISCUSSION THREAD

**Discussion by cvcore** [Homebrew/discussions#3908] [Writing Formulae/Casks] [answered]
Greetings!

I noticed that the formula `sshfs` has been disabled due to the depency on macFUSE. But I was able to download the source code from https://github.com/osxfuse/sshfs/releases/tag/osxfuse-sshfs-2.5.0 and do a manual compile and install without big issue (I only need to remove the `#include <fuse_darwin.h>` from `sshfs.c`).

However when I try to package it into a formula `osxfuse-sshfs`, the `make` command no longer works, giving me an error when I run `brew install --build-from-source osxfuse-sshfs`:
```
...
==> make
Last 15 lines from /Users/***/Library/Logs/Homebrew/osxfuse-sshfs/03.make:
2022-11-12 15:39:14 +0000

make

/Applications/Xcode.app/Contents/Developer/usr/bin/make  all-am
clang -DHAVE_CONFIG_H -I.  -D_REENTRANT -DFUSE_USE_VERSION=26 -DLIBDIR=\"/opt/homebrew/Cellar/osxfuse-sshfs/2.5.0_1/lib\"  -D_FILE_OFFSET_BITS=64 -I/usr/local/include/fuse -I/opt/homebrew/Cellar/pcre2/10.40/include -I/opt/homebrew/Cellar/glib/2.74.0/include -I/opt/homebrew/Cellar/glib/2.74.0/include/glib-2.0 -I/opt/homebrew/Cellar/glib/2.74.0/lib/glib-2.0/include -I/opt/homebrew/opt/gettext/include -g -O2 -Wall -W -Icompat -c -o sshfs-sshfs.o `test -f 'sshfs.c' || echo './'`sshfs.c
sshfs.c:12:10: fatal error: 'fuse.h' file not found
#include <fuse.h>
         ^~~~~~~~
1 error generated.
make[1]: *** [sshfs-sshfs.o] Error 1
```

The strange thing is, when I enter the interactive shell after failure, I was able to build the source code without issue with
`/usr/bin/make` but never with `make` (which resolves to `/opt/homebrew/Library/Homebrew/shims/mac/super/make`).

So for the time being I changed the make invocation to
```
system "/usr/bin/make"
system "/usr/bin/make", "install"
# instead of
# system "make"
# system "make", "install"
```
to make it work, but I believe this is not the clean way to do it.

**What I would like to know:**
1. What is the reason of this awkward behavior of `make`?
2. How to write the brew formula correctly in this case?

Thanks!

--------------

Here is the formula I created:

```ruby
class OsxfuseSshfs < Formula
  desc "File system client based on SSH File Transfer Protocol"
  homepage "https://github.com/osxfuse/sshfs"
  url "https://github.com/osxfuse/sshfs/archive/osxfuse-sshfs-2.5.0.tar.gz"
  sha256 "8ea4d3d5bc0f343998009d7eb138e3804490f6a22e890c6de4eadc6fd2414ae0"
  license any_of: ["LGPL-2.1-only", "GPL-2.0-only"]
  revision 1

  depends_on "autoconf" => :build
  depends_on "automake" => :build
  # depends_on "libtool" => :build
  # depends_on "pkg-config" => :build
  depends_on "glib"
  depends_on :macos


  patch :DATA

  def install
    ENV.deparallelize  # if your formula fails when building in parallel

    system "autoreconf", "-i"
    system "./configure", *std_configure_args
    system "/usr/bin/make"
    system "/usr/bin/make", "install"
  end

  test do
    system "#{bin}/sshfs", "--version"
  end
end
__END__
diff --git a/sshfs.c b/sshfs.c
index b09484c..e90382e 100644
--- a/sshfs.c
+++ b/sshfs.c
@@ -13,7 +13,7 @@
 #include <fuse_opt.h>
 #include <fuse_lowlevel.h>
 #ifdef __APPLE__
-#  include <fuse_darwin.h>
+//#  include <fuse_darwin.h>
 #endif
 #include <assert.h>
 #include <stdio.h>
```

**Comment by gromgit** [comment, marked as answer]
> * What is the reason of this awkward behavior of `make`?

Homebrew build shims sanitize their environment to ensure reproducible builds. One step involves dropping `/usr/local` as a source of include files and libraries for M1 Macs. Guess where MacFUSE installs its include files and libraries?

> * How to write the brew formula correctly in this case?

You could look at how it's done in my FUSE tap: https://github.com/gromgit/homebrew-fuse/. In short, it involves copying the MacFUSE files to a temp dir and setting various environment variables during the build process.

Or you could just `brew install gromgit/fuse/sshfs-mac`. :grinning:

**Reply by cvcore** [reply]
That answers exactly my question and thanks for maintaining the formula.

Cheers!
